### ISSUE TITLE
Error creating dev database - could not ping postgres on startup: reached retry deadline

### ISSUE THREAD

**Issue by matt-richardson** [hashicorp/boundary#1366]
**Describe the bug**

Running `boundary dev` on windows fails to launch postgres with error
```
Error creating dev database container unable to start dev database with dialect postgres: could not ping postgres on startup: reached retry deadline
```

**To Reproduce**
1. Run `boundary dev`
2. See error

**Expected behavior**
First start demo experience as described [here](https://learn.hashicorp.com/tutorials/boundary/getting-started-dev?in=boundary/getting-started) should be seamless, not painful.

**Additional context**
This is my first time looking at boundary - so far it's been pretty rocky. This does not bode well.

My journey looks something like:
* start reading [docs](https://learn.hashicorp.com/collections/boundary/getting-started).
* get to [install section](https://learn.hashicorp.com/tutorials/boundary/getting-started-install?in=boundary/getting-started) - instead of a download link, it recommends using some unknown third party package manager that I've barely heard of. Not awesome for a security tool.
* go find the direct download page. download it. It's just a zip file - where do I put it? Why isn't there an installer? Does it need to be on the path? Will I get bitten if I install it under `C:\Program Files`? Does it need write access to it's directory for logs etc?
* try running `boundary dev`
* get an error `Error creating dev database container: unable to start dev database with dialect postgres: could not start resource: : Post "http://localhost:2375/images/create?fromImage=postgres&tag=12": dial tcp [::1]:2375: connectex: No connection could be made because the target machine actively refused it.`
* find out that it's trying to talk to docker over an unsecure port, and there's instructions on the "learn" page on how to disable TLS. On a tool that prides itself on taking security seriously. Really?
* okay, acknowledge the inevitable and expose docker on 2375. Find that docker cant access it. Go searching for a while and eventually [find](https://github.com/docker/for-win/issues/3546#issuecomment-586101121) I have to run `netsh int ipv4 add excludedportrange protocol=tcp startport=2375 numberofports=1`. Reboot.
* try running `boundary dev` again. get the error `Error creating dev database container unable to start dev database with dialect postgres: could not ping postgres on startup: reached retry deadline`.
* look at docker - okay, it's launching the docker container, postgres is running, it's sitting there waiting for connections happily.
* Search issues for the error message. no luck.
* try adding `--log-level debug` to see what's happening. No logs.
* try `--log-level trace` to see what's happening. No logs.
* try `--log-level trace --log-format standard` to see what's happening. No logs.
* try searching for issues again. Find a vaguely related issue and end up trying [this solution](https://github.com/hashicorp/boundary/issues/720#issuecomment-712472029) which is basically "run your own db"
* realise that the suggested way of talking to the database disables ssl mode. Really?
* start my own copy of postgres (trying to guess what params i need)
* launch boundary pointing at this copy of the database
* and finally it launches.

I'm sharing this as I know you've put a lot of effort into making the onboarding smooth and easy to try, but unfortunately, for me it's been a not very fun journey so far.

**Comment by matt-richardson** [comment]
To workaround this, i ran:
```
docker run -e POSTGRES_PASSWORD=passwordgoeshere -e POSTGRES_DB=boundary -p 5432:5432 postgres:11 docker-entrypoint.sh postgres
```
and then
```
boundary dev --log-level trace --log-format standard -database-url=postgres://postgres:passwordgoeshere @127.0.0.1:5432?sslmode=disable
```

**Comment by malnick** [comment]
Thanks for submitting this @matt-richardson, and sorry to hear you're having a rough time getting started with Boundary. I'll do my best to answer your bullets inline here: 

> get to install section - instead of a download link, it recommends using some unknown third party package manager that I've barely heard of. Not awesome for a security tool.

There are a limited number of developer-based package managers for Windows, and we felt [scoop](https://scoop.sh/) was a really good choice given the ease of use and overall workflow. 

> go find the direct download page. download it. It's just a zip file - where do I put it? Why isn't there an installer? Does it need to be on the path? Will I get bitten if I install it under C:\Program Files? Does it need write access to it's directory for logs etc?

If you install via `scoop` an installer script will run and the binary will be available in your `PATH`. However, and we did our best to point this out on the [learn page](https://learn.hashicorp.com/tutorials/boundary/getting-started-install?in=boundary/getting-started), there is an option to download and install it manually. This route assumes the operator wants the freedom to install where they wish. Yes, the binary needs to be somewhere in your `PATH`, and specifically where this is can differ from operator to operator. If you're unsure, I recommend using `scoop` which will take care of this for you. 

We do realize there's a gap in our documentation on server-side installations since we only cover using [Linux `systemd` installs](https://www.boundaryproject.io/docs/installing/systemd). We haven't had a request for server-side Windows installs and would love to hear your thoughts on adding this if it's something you're considering. 

> it's trying to talk to docker over an unsecure port

The Docker client will default to connecting to unix:///var/run/docker.sock on Linux, and tcp://127.0.0.1:2376 on Windows. There are some [good getting started guides here](https://docs.docker.com/engine/reference/commandline/dockerd).

> look at docker - okay, it's launching the docker container, postgres is running, it's sitting there waiting for connections happily.

Can you tell us more about your Window's docker setup? Are you using docker desktop for windows? As I understand it, this runs docker in a VM on Windows and there are some [caveats with networking there](https://docs.docker.com/docker-for-windows/networking/) that we might be able to find some good work arounds for with more info at hand. 

> disabling TLS and other SSL caveats when running `boundary dev`

[dev mode](https://www.boundaryproject.io/docs/getting-started#what-is-dev-mode) is meant to be the easiest way to get started with Boundary. We run the database as a container using Docker and disable all TLS capabilities by default to avoid the complexities of managing PKI infrastructure. Being short for "Development" mode, this mode of operation is by default insecure, and is intended to help our community develop against a running Boundary environment, get started with basic workflows, and isn't intended to be used for any production-like workloads. 

Along with disabling TLS we also generate Boundary resources like login accounts, authentication methods, targets and even a dev OIDC server for testing our OIDC authentication method against. If you're interested in running Boundary in a production environment, check out our [docs on non-dev environments](https://www.boundaryproject.io/docs/installing/no-gen-resources) which covers what basic resources need to be established to get running outside of development mode.

**Comment by jefferai** [comment]
Assuming you're running on WSL, you are almost certainly hitting this longstanding, ongoing Docker-for-Windows issue: https://github.com/docker/for-win/issues/3171#issuecomment-788808021

One of the really frustrating things about it is Docker doesn't _tell_ you what the actual problem is (that it is failing to bind the port for security reasons). I've spent hours tracking this down and finding that workaround.

For reasons I have not yet figured out, when running the container manually it uses a different port range than when it is run via Boundary dev, and this generally seems to avoid the problem. But the root issue is that Docker is failing to allocate a port, then going along its merry way instead of allocating a different one, along with the fact that it is using different port ranges based on.....something.

The specific comment I linked to above is what has worked for me. I should note that the first call (to stop winnat) sometimes hangs so you need to Ctrl-C it. You'll need to reboot after this. And the problem _will_ come back now and again. :-(

**Comment by DerekStrickland** [comment]
FWIW, I ran into similar error messages as well, though I am running on macOS. My workaround was that I had to run the Postgres container once manually, with the following command. 

```
docker run -e POSTGRES_PASSWORD=password -e POSTGRES_DB=boundary -p 5432:5432 postgres:11 docker-entrypoint.sh postgres
``` 

Note I was in the `docker` directory of the repo. If you aren't, you'll need to qualify the path to the `docker-entrypoint.sh` file.

After I manually started the Postgres container, `boundary dev` worked. I then stopped both the `boundary dev` process and the running Postgres container. Now I am able to restart `boundary dev` as many times as I like without having to manually start the Postgres container. 

I did some step through of the code, and I think what I saw was that [`docker.startDbInDockerSupported`](https://github.com/hashicorp/boundary/blob/86245c48da984a5034e735d41ba843466209ffde/internal/cmd/base/internal/docker/supported.go#L20) finished without error, but then code that tries to dial that Postgres URL fails, and then you see a network error. I immediately questioned whether the container actually started, and  `docker ps` verified the container never started. It might also have been the call to `common.SqlOpen` within `docker.startDbInDockerSupported`. Apologies for not recalling. It was yesterday. 

After a quick issue search, I found this one which inspired me to try running the container manually as a workaround. I figured I'd add these details, and share that I had issues on macOS, so it's possible to run into issues getting started on macOS as well.  Perhaps this is the expected workflow, and the tutorial needs an update, but I'm guessing not? I was following [Start a Development Environment tutorial](https://learn.hashicorp.com/tutorials/boundary/getting-started-dev?in=boundary/getting-started). FWIW, it has a section on troubleshooting DB container creation on Windows, but the referenced setting is not exposed in Docker Desktop for Mac. I interpreted that to mean isn't relevant to macOS users.
